### PR TITLE
Lps 173758 [draft proposal] clay tabs tag

### PR DESCRIPTION
**Pull request by dsanz**
In this PR I'm proposing a mechanism to mimick clay tags composition in the browser side when it comes to hydrate clay react composable components. Mechanism basically:
* Defers hydration of react components when the container div is not in the page
* Provides a way for composable react components to trigger the hydration for those markup they generate

Surprisingly, we did not need this before, as most od the clay composable tags don't use the react equivalent.

We discovered this while working on https://issues.liferay.com/browse/LPS-174606 in the clay tabs tab. I sent the pull on top of the initial WIP stuff Carlos started for the tag. Actual pull for the tag is https://github.com/liferay-frontend/liferay-portal/pull/3020

This is a draft pr to discuss the approach with @carloslancha, @markocikos and @kresimir-coko 

I'm also ccing @matuzalemsteles and @ethib137 fyi

### CODE REVIEW

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/Tabs.js [hunk 1]
```diff
@@ -37,6 +37,27 @@ export default function Tabs({
...
+    }
+	  return null;
+  }
+
+
+	useEffect(() => {
+    panels.map((panel) => {
+        const panelDocument = new DOMParser().parseFromString(panel, "text/html");
+        panelDocument.body.querySelectorAll("div div[id^=__RENDER__]").forEach((c) => {
+          const id = isRenderNode(c);
```
Comment (dsanz) [inline modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/Tabs.js:55]: when inline rendering is on, `panel` will not only contain the container markup for hydration but also the script that calls render.
As Tabs is using `dangerouslySetInnerHTML` for the panels, these scripts [won't run ](https://suncommander.medium.com/how-to-inject-scripts-using-dangerouslysetinnerhtml-on-client-side-rendering-973037cc06b7).

So, this effect can also be in charge of handling the execution of these as they can be easily detected

Comment (dsanz) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/Tabs.js:55]: another caveat is that we probably don't want to run the hydration for all the potential nested tags but only to the first-level ones, in order to guarantee proper hydration order in the hierarchy. This is something I have not checked

**Review by matuzalemsteles — COMMENTED** [review]
This looks like an interesting solution to handle composition and hydration of components with taglibs, I think for simpler cases this might work well.

I think part of this work we can explore other ideas to deal with nested composition and more complex cases if we want to offer that as well. For example this will not work well if you have contexts in React, nested rendering are different React trees, a state change does not cause a rendering in components in different trees, also as we will have to defer nested hydration this can be slow and this affects the [Web Vitals FID (First Input Delay)](https://web.dev/fid/) metric, for example:

- We will have to wait for the browser to analyze the JS and execute
- React rendering
- Wait for React to call lifecycles
- Parse the HTML
- Invoke component rendering again

I'm thinking of other possibilities to allow us to hydrate nested components and all share the same React tree without having to deal with lifecycles, html parsing and deferring renders.

We can create a mechanism on the taglibs side that can simulate creating a React tree. For example, when rendering a Button component inside a Tabs taglib, the Button taglib generates a JSON object with the structure of the JSX object serilized in a script tag and sends it along with the initial rendering, the output could be like this:

```html
<script>
   self.__liferay_hydrate.push([parentId, component, props]);
</script>
```

We need to ensure that `self.__liferay_hydrate = []` is declared initially and first in all scripts to ensure order of execution.

Where there is `component` can be the string of the HTML tag or the reference to the React.js component, for the reference maybe we would need the bundler to load the component and add the component reference to the array.

The tabs component being the root component would not need to generate the object, the `parentId` is to help mount the tree before React rendering, mount the JSON structure and convert the object into a JSX element with the help of [`React.createElement`](https://beta.reactjs.org/learn/add-react-to-a-website#react-without-jsx) and pass it as `children` for the rendering of the Tabs component, this step can be done in the `render` method.

If the component has elements that are strings they are also passed to the synthetic jsx json, we can create the same structure that [`React.createElement` expects `React.createElement(component, props, ...children)`](https://beta.reactjs.org/learn/add-react-to-a-website#react-without-jsx).

For example, considering the Tabs scenario, with a children:

```html
// taglib 
<clay:tabs
  tabsItems="<%= tabsDisplayContext.getDefaultTabsItems() %>"
>
  <clay:tabs-panel>
    Tab Content 2
    <clay:button propsTransformer='js/ClaySampleButtonPropsTransformer'>Two</clay:button>
  </clay:tabs-panel>
</clay:tabs>

// Output
<script>
  self.__liferay_hydrate.push(['tabsId', ClayTabs.TabPanel, null, 'Tab Conent 2']);
</script>

<script>
  self.__liferay_hydrate.push(['tabsId', ClayButton, null, 'Two']);
</script>

// Later

React.createElement(ClayTabs.TabPanel, null, 'Tab Conent 2', React.createElement(ClayButton, null, Two))
```

The scenario here would be to know if we can find out via taglib if there is a taglib parent to get the id of the instance to make a linked list. Another scenario is that we could do the same for pure html content instead of passing a string and using `dangerouslySetInnerHTML`.

If we can implement this, taglibs like the `clay:button` component when it's nested within another component, it just generates the HTML markup on the server and the object, it wouldn't be necessary to call the `render` method.

- Nested components are rendered within the same React tree
- We can take advantage of using nested React context
- State update affects entire component tree
- We don't need to parse the html
- We don't need to deferring the rendering, everything will be done by React itself, as if it were a full React application.
- We don't need to implement hooks or extra logic inside the React components to deal with the HTML parse and invoke the render methods, the implementation is simpler.

**Comment by carloslancha** [comment]
Dropping here another poc with a different approach https://github.com/liferay-frontend/liferay-portal/pull/3045

**Comment by dsanz** [comment]
That's an interesting proposal @matuzalemsteles , thanks for taking the time to explain it in detail. One of the things that crossed my mind while working on this is whether we shall keep improving the rendering infra to fully support tag nesting or not. The thing is that I feel this mechanism will promote the usage of taglibs instead of react, ant that's not good imho as it is not promoting adoption of React. Said that, we need to support these tags for some time and even if we moved our codebase to pure React, some DXP context is still needed.

It's true that at this moment, each hydrated component is isolated in terms of react tree. They don't share anything but the general context we inject in each render operation. Any communication between react parts must be done via taglib attributes (shared JSP display context), or else making a bigger react component and rendering it altogether. 

Said that, I'm a bit concerned that a solution which allows JSP devs to use React via tags "seamlessly", because in that case we could risk increasing the devs expectations about what clay tags can do. For example, if there is a clay component which defines context and that context is needed in a child component, it's probably much better to encapsulate the whole thing in a single tag or ask developers to create a composed React solution, rather than trying to "keep the parity" and allow tag nesting. 

Still, we need some sort of solution supporting composability for certain cases. My first attempt was made from the bottom and considering the way markup was being injected (`innerHTML`) which is probably something that shall be changed. The route I'd like to explore is enabling this for the tags that really need it, namely:
* Composable ones and
* Produce SSR + React hydration on its own
* Accept another clay tag in the bidy that needs react hydration

Does that make sense to you?

cc @carloslancha

**Comment by dsanz** [comment]
To centralize discussion I'm closing this one in favor of https://github.com/liferay-frontend/liferay-portal/pull/3045, please feel free to keep discussing there

**Pull request closed without merging**